### PR TITLE
[BE] Project manager role

### PR DESCRIPTION
**Pull request by nspc97**
[BE] Project manager role
https://trello.com/c/mY9KaA6l/99-be-project-manager-role

- Each user should get a PM assigned (not mandatory).
- Admins can assign a PM to the user.
- When the PM account is deleted/deactivated, all the users of this PM get unassigned from him.

### CODE REVIEW

File: src/Controller/UserController.php [hunk 1]
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
...
+            ], Response::HTTP_FORBIDDEN);
+        }
+        $user->setRoles(array('ROLE_PM'));
+        $userRepository->save($user);
+        return new JsonResponse(['message' => 'Role successfully set up!'], Response::HTTP_OK);
+    }
+
+    /**
+     * Set Project Manager
+     * @Rest\Post("/{userId}/set_pm/{pmId}", requirements={"id"="\d+"})
```
Comment (MihaiDruta) [inline src/Controller/UserController.php]: Requirements are not correct

File: src/Controller/UserController.php [hunk 2]
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
...
+     *     @SWG\Property(property="code", type="integer", example=404),
+     *     @SWG\Property(property="message", type="string", example="Not found!"),
+     *     )
+     * )
+     * @SWG\Response(
+     *     response="400",
+     *     description="This user in not a PM!",
+     *     @SWG\Schema(
+     *     @SWG\Property(property="code", type="integer", example=400),
+     *     @SWG\Property(property="message", type="string"),
```
Comment (MihaiDruta) [inline src/Controller/UserController.php]: You can add an example message for this response

File: src/Entity/User.php [hunk 3]
```diff
@@ -160,8 +160,15 @@ class User implements UserInterface
      * @Groups({"UserDetail"})
      * @SWG\Property(example="15555555599")
      */
+
     private $createdAt;
 
+    /**
+     * Comment id linked to.
```
Comment (MihaiDruta) [inline src/Entity/User.php]: Wrong description

**Review by MihaiDruta — CHANGES_REQUESTED** [review]

**Review by MihaiDruta — APPROVED** [review]

File: src/Repository/UserRepository.php [hunk 4]
```diff
@@ -60,6 +60,7 @@ public function save(User $newUser): void
     public function delete(User $user): void
     {
         $this->commentRepository->anonymizeUserComments($user);
+        $this->clearProjectManager($user);
```
Comment (SpartakusMd) [inline src/Repository/UserRepository.php]: Rename to `unassignProjectManager`

File: src/Controller/UserController.php [hunk 5]
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
...
+            ], Response::HTTP_FORBIDDEN);
+        }
+        $user->setRoles(array('ROLE_PM'));
+        $userRepository->save($user);
+        return new JsonResponse(['message' => 'Role successfully set up!'], Response::HTTP_OK);
+    }
+
+    /**
+     * Set Project Manager
+     * @Rest\Post("/{userId}/set_pm/{pmId}", requirements={"userId"="\d+", "pmId"="\d+"})
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: Rename `set_pm` to `assign-project-manager`. The ID of PM should be present in the body not in the URL.

File: src/Controller/UserController.php [hunk 6]
```diff
@@ -610,4 +611,164 @@ public function removeAvatar(
         }
         return new JsonResponse(['message' => 'Avatar successfully deleted!'], Response::HTTP_OK);
     }
+
+    /**
+     * Set Project Manager role
+     * @Rest\Post("/{id}/set_pm", requirements={"id"="\d+"})
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: Role management should be general. Only admins can assign roles. An admin can promote anyone to PM or to admin and also demote to PM, user.

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

File: src/Controller/UserController.php [hunk 7]
```diff
@@ -613,21 +613,28 @@ public function removeAvatar(
     }
 
     /**
-     * Set Project Manager role
-     * @Rest\Post("/{id}/set_pm", requirements={"id"="\d+"})
+     * Set role
+     * @Rest\Post("/{id}/set_role", requirements={"id"="\d+"})
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: In URLs dash "-" should be used instead of underscore "_".

File: src/Controller/UserController.php [hunk 8]
```diff
@@ -662,29 +669,54 @@ public function removeAvatar(
...
         $userRepository->save($user);
         return new JsonResponse(['message' => 'Role successfully set up!'], Response::HTTP_OK);
     }
 
     /**
      * Set Project Manager
-     * @Rest\Post("/{userId}/set_pm/{pmId}", requirements={"userId"="\d+", "pmId"="\d+"})
-     * @ParamConverter("user", options={"mapping": {"userId" : "id"}})
-     * @ParamConverter("projectManager", options={"mapping": {"pmId" : "id"}})
+     * @Rest\Post("/{id}/assign_project_manager", requirements={"id"="\d+"})
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: In URLs dash "-" should be used instead of underscore "_".

File: src/DTO/UserDTO.php [hunk 9]
```diff
@@ -187,6 +187,16 @@ class UserDTO
      */
     public $avatar;
 
+    /**
+     * User role
+     * @var string
+     * @Serializer\Type("string")
```
Comment (SpartakusMd) [inline src/DTO/UserDTO.php]: A user can have multiple roles.

File: src/DTO/UserDTO.php [hunk 10]
```diff
@@ -197,4 +207,15 @@ public function isLocationValid(ExecutionContext $context): void
...
     }
+
+    /**
+     * @param ExecutionContext $context
+     * @Assert\Callback(groups={"UserRole"})
+     */
+    public function isRoleValid(ExecutionContext $context): void
+    {
+        if (!in_array($this->roles, User::ROLES, true)) {
+            $context->addViolation('Please enter a valid ROLE!', array());
```
Comment (SpartakusMd) [inline src/DTO/UserDTO.php]: The field for which this violation is, was not specified.

File: src/Controller/UserController.php [hunk 11]
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
                 'message' => 'Access denied!'
             ], Response::HTTP_FORBIDDEN);
         }
+
+        $data = $request->getContent();
+        /** @var DeserializationContext $context */
+        $context = DeserializationContext::create()->setGroups(array('SetRole'));
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: Why is the group called `SetRole`?

File: src/Controller/UserController.php [hunk 12]
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
...
+            return new JsonResponse(
+                [
+                    'code' => Response::HTTP_NOT_FOUND,
+                    'message' => 'Not found!'
+                ],
+                Response::HTTP_NOT_FOUND
+            );
+        }
+
         if ($projectManager->getRoles() === array('ROLE_USER')) {
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: This isn't how you validate if a user has a specific role. Also, you should check specificaly for ROLE_PM as there could be roles like DEV, QA and others.

File: src/Controller/UserController.php [hunk 13]
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
...
-            return new JsonResponse(['message' => 'This user in not a PM!'], Response::HTTP_BAD_REQUEST);
+            return new JsonResponse(
+                [
+                    'code' => Response::HTTP_BAD_REQUEST,
+                    'message' => 'This user in not a PM!'
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: The message is bad. It says that the user to be changed is not a PM, not that the current user is not a PM.

File: src/Controller/UserController.php [hunk 14]
```diff
@@ -764,8 +796,36 @@ public function setProjectManagerForUser(
...
+            User::class,
+            'json',
+            $context
+        );
+        $projectManager = $userRepository->find($id);
+        if (!$projectManager) {
+            return new JsonResponse(
+                [
+                    'code' => Response::HTTP_NOT_FOUND,
+                    'message' => 'Not found!'
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: What is not found?

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

File: src/Controller/UserController.php [hunk 15]
```diff
@@ -812,17 +814,17 @@ public function setProjectManagerForUser(
...
-                    'message' => 'Not found!'
+                    'message' => 'Project Manager to assign not found!'
                 ],
                 Response::HTTP_NOT_FOUND
             );
         }
-
-        if ($projectManager->getRoles() === array('ROLE_USER')) {
+        $projectManagerRole = $projectManager->getRoles();
+        if ($projectManagerRole !== array('ROLE_PM') || $projectManagerRole !== array('ROLE_ADMIN')) {
```
Comment (SpartakusMd) [inline src/Controller/UserController.php]: You should be using a specific Symfony function to check the role. As ROLE_ADMIN ha already ROLE_PM you won't have to specify it.

**Review by SpartakusMd — CHANGES_REQUESTED** [review]

**Pull request merged**